### ISSUE TITLE
DL3059 rule does not detect RUN instructions separated by comments or line carries

### ISSUE THREAD

**Issue by amber-beasley-liatrio** [hadolint/hadolint#1138]
# DL3059 rule should align with SonarQube's detection of RUN instructions separated by comments or other instructions

## Summary

The DL3059 rule "Multiple consecutive RUN instructions. Consider consolidation." only detects RUN instructions that are literally consecutive (adjacent lines) but does not catch patterns where RUN instructions could be consolidated but are separated by comments or other Dockerfile instructions. This behavior is inconsistent with similar static analysis tools like SonarQube, which correctly identifies RUN instructions that could be consolidated even when separated by comments or non-critical instructions.

## Expected Behavior

DL3059 should detect RUN instructions that could potentially be consolidated, even when separated by:

- Comments
- Non-critical instructions (like WORKDIR when it doesn't change the context significantly)

## Current Behavior

DL3059 only triggers on truly consecutive RUN instructions with no intervening lines.

### Case 1: Comments between RUN instructions (NOT detected, but arguably should be)

```dockerfile
RUN groupadd --gid 1001 builduser \
    && useradd --uid 1001 --gid builduser --shell /bin/bash --create-home builduser

# Install dependencies with security hardening
RUN apt-get update \
    && apt-get install -y --no-install-recommends \
        curl \
        gdb \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/*

# Set working directory and ownership  
WORKDIR /app
RUN chown builduser:builduser /app
```

### Case 2: Truly consecutive RUN instructions (correctly detected)

```dockerfile
RUN groupadd --gid 1001 builduser
RUN apt-get update  # ← DL3059 triggered
RUN chown builduser:builduser /app  # ← DL3059 triggered
```

## Rationale for Enhancement

1. **Dockerfile optimization**: The separated RUN instructions in Case 1 could often be consolidated for better layer efficiency
2. **Consistency with industry tools**: SonarQube's Docker analysis correctly identifies consolidatable RUN instructions even when separated by comments, making hadolint's detection less comprehensive
3. **Tool alignment**: Teams using both hadolint and SonarQube get inconsistent feedback, with SonarQube catching optimization opportunities that hadolint misses
4. **Developer education**: Many developers separate RUN instructions with comments thinking it's better practice, but miss optimization opportunities that other linters would catch

### Option 1: Configurable Detection Scope

Add configuration to control what separates "consecutive" RUN instructions:

```yaml
# .hadolint.yaml
override:
  DL3059:
    ignore_comments: false  # Consider RUN instructions separated only by comments as consecutive
    ignore_workdir: true    # Don't consider WORKDIR as breaking consecutive RUN pattern
```

### Option 2: New Rule Variant

Create a new rule (e.g., DL3060) for "potentially consolidatable RUN instructions" with broader detection.

### Option 3: Enhanced DL3059

Extend DL3059 to have different severity levels or sub-rules for different separation patterns.

## Environment

- hadolint version: 2.12.0
- Platform: macOS
- Dockerfile complexity: Multi-stage builds with security hardening

## Additional Context

This came up during implementation of strict Dockerfile linting with `failure-threshold: info` where we expected DL3059 to catch optimization opportunities that it currently misses, similar to how SonarQube's Docker analysis works.

**SonarQube Comparison**: SonarQube's rule "Docker instructions should be grouped when it makes sense" (rule key: `docker:S6473`) successfully detects RUN instructions that could be consolidated even when separated by:

- Comments
- Non-critical instructions like WORKDIR
- Blank lines

This makes SonarQube more effective at identifying Dockerfile optimization opportunities compared to hadolint's current implementation.

The current behavior is technically correct for "consecutive" instructions, but aligning with SonarQube's broader detection would make hadolint more comprehensive and consistent with industry standards.
